### PR TITLE
Refactored item serializer to better encompass data validation. Added…

### PR DESCRIPTION
**Pull request by brodykellish**
… custom field data to item creation/modification endpoints

### CODE REVIEW

File: kipventory/api/serializers.py [hunk 1]
```diff
@@ -198,15 +254,15 @@ class Meta:
...
+def validate_username(instance, value):
+    netid_regex = re.compile(r'[a-z]{2,3}[0-9]{1,3}')
+    if netid_regex.fullmatch(value):
+        raise serializers.ValidationError("Username cannot be the same form as Duke NetID.")
     username_taken = (User.objects.filter(username=value).count() > 0)
     if username_taken:
-        raise serializers.ValidationError("Username is already taken.")
+        if instance is not None:
+            if not (instance.username == value):
+                raise ValidationError({"username": ["Username '{}' is already taken.".format(username)]})
```
Comment (jes85) [inline kipventory/api/serializers.py]: question about instance

File: kipventory/api/urls.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
     url(r'^items/(?P<item_name>[\w\s]+)/?$',                              views.ItemDetailModifyDelete.as_view()),
     url(r'^items/(?P<item_name>[\w\s]+)/addtocart/?$',                    views.AddItemToCart.as_view()),
     url(r'^items/(?P<item_name>[\w\s]+)/fields/?$',                       views.CustomValueList.as_view()),
-    url(r'^items/(?P<item_name>[\w\s]+)/fields/(?P<field_name>[\w\s]*)/?$', views.CustomValueDetailModify.as_view()),
+    # url(r'^items/(?P<item_name>[\w\s]+)/fields/(?P<field_name>[\w\s]*)/?$', views.CustomValueDetailModify.as_view()),
```
Comment (jes85) [inline kipventory/api/urls.py]: ?

File: kipventory/api/views.py [hunk 3]
```diff
@@ -57,16 +57,7 @@ def get_queryset(self):
     def get_serializer_class(self):
         return serializers.ItemSerializer
 
-    def get(self, request, format=None):
```
Comment (jes85) [inline kipventory/api/views.py:60]: why so many deletions?

File: kipventory/api/serializers.py [hunk 4]
```diff
@@ -291,20 +340,15 @@ class Meta:
...
         status = data.get('status', None)
 
-        return {
-            "date_closed": date_closed,
-            "closed_comment": closed_comment,
-            "administrator": administrator,
-            "status": status
-        }
+        validated_data = super(RequestPUTSerializer, self).to_internal_value(data)
+        validated_data.update({"date_closed": date_closed, "administrator": administrator})
```
Comment (jes85) [inline kipventory/api/serializers.py]: .update()?

File: kipventory/api/serializers.py [hunk 5]
```diff
@@ -3,7 +3,8 @@
 from . import models
 from django.contrib.auth.models import User
 from django.utils import timezone
-import re
```
Comment (jes85) [inline kipventory/api/serializers.py:6]: I should review all this stuff in more detail later for my understanding

File: assets/js/app/manage/customfields/CustomFieldContainer.jsx [hunk 6]
```diff
@@ -243,10 +234,10 @@ const CustomFieldContainer = React.createClass({
                 </Col>
                 <Col sm={6}>
                   <FormControl componentClass="select" name="field_type" value={this.state.field_type} onChange={this.onChange}>
-                    <option value="s">Short Text</option>
-                    <option value="m">Long Text</option>
-                    <option value="i">Integer</option>
-                    <option value="f">Float</option>
+                    <option value="Single">Single-line text</option>
```
Comment (jes85) [inline assets/js/app/manage/customfields/CustomFieldContainer.jsx:237]: Why name change?

File: assets/js/app/inventory/ItemModificationModal.jsx [hunk 7]
```diff
@@ -4,7 +4,7 @@ import { FormGroup, Button, Modal, FormControl, ControlLabel }  from 'react-boot
 import { getCookie } from '../../csrf/DjangoCSRFToken'
```
Comment (jes85) [inline assets/js/app/inventory/ItemModificationModal.jsx:4]: I'm trusting you on front end stuff but I should review it for my understanding later

File: assets/js/app/inventory/ItemDetail.jsx [hunk 8]
```diff
@@ -399,13 +396,19 @@ const ItemDetail = React.createClass({
...
       }
 
-      state.tags = tagArray;
 
       ajax({
         url:"/api/items/" + thisobj.state.item.name + "/",
         type: "PUT",
         traditional: true,
-        data: state,
+        data: {
```
Comment (jes85) [inline assets/js/app/inventory/ItemDetail.jsx:403]: we should standardize how we send json. I think stringifying the whole thing is better

File: assets/js/app/KipNav.jsx [hunk 9]
```diff
@@ -13,54 +13,69 @@ const KipNav = React.createClass({
...
             </Navbar.Brand>
             <Navbar.Toggle />
           </Navbar.Header>
           <Navbar.Collapse>
             <Nav>
-              <LinkContainer to="/app/requests/">
-                <NavItem eventKey={1}>Requests</NavItem>
-              </LinkContainer>
+              {this.getInventoryLink()}
+              {this.getLink("/app/requests/", 1, "Your Requests")}
```
Comment (jes85) [inline assets/js/app/KipNav.jsx]: change event keys back?

File: assets/js/app/KipNav.jsx [hunk 10]
```diff
@@ -13,54 +13,69 @@ const KipNav = React.createClass({
...
+              <Link to="/app/inventory/">kip-ventory</Link>
             </Navbar.Brand>
             <Navbar.Toggle />
           </Navbar.Header>
           <Navbar.Collapse>
             <Nav>
-              <LinkContainer to="/app/requests/">
-                <NavItem eventKey={1}>Requests</NavItem>
-              </LinkContainer>
+              {this.getInventoryLink()}
```
Comment (jes85) [inline assets/js/app/KipNav.jsx]: use getLink method

File: kipventory/api/serializers.py [hunk 11]
```diff
@@ -187,6 +191,11 @@ def update(self, instance, validated_data):
         item_data = validated_data
         # Create the item instance
         item = super(ItemSerializer, self).update(instance, item_data)
+
+        if item_data.get('tags', None) is None:
```
Comment (jes85) [inline kipventory/api/serializers.py]: if tags is not included should we delete tags? what if update is partial

File: assets/js/app/inventory/ItemDetail.jsx [hunk 12]
```diff
@@ -396,6 +396,7 @@ const ItemDetail = React.createClass({
         var tagArray = tags;
       }
 
+      console.log(tagArray)
```
Comment (jes85) [inline assets/js/app/inventory/ItemDetail.jsx]: delete prints

File: kipventory/api/serializers.py [hunk 13]
```diff
@@ -80,61 +85,119 @@ def is_item_in_cart(self, item):
...
-        item_values = item.values.all()
-        # If we have
-        for field, value in field_data.items():
-            try:
-                cv = item_values.get(field__pk=field.pk)
-                setattr(cv, field.field_type, value)
-                cv.save()
-            except:
-                print("baseee")
+        if item_data.get('tags', None) is None:
```
Comment (jes85) [inline kipventory/api/serializers.py:190]: if tags is not included should we delete tags? what if update is partial

**Comment by brodykellish** [comment]
I'll fix these issues and integrate them with the new CartItem/RequestedItem models that I've been working on.

**Pull request closed without merging**